File: 29-04/eperfeito.py
def func_mod(a, b):
    if a == b:
        return 0
    elif a > b:
        return func_mod(a - b, b)
    elif a < b:
        return a

def soma_div(n, d, s):
    if d == 0:
        return s
    elif func_mod(n, d) == 0:
        return soma_div(n, d-1, s + d)
    else:
        return soma_div(n, d-1, s)

def eperfeito(n):
    if soma_div(n, n-1, 0) == n:
        return True
    else:
        return False



if __name__ == "__main__":
    print("Digite um número:")
    print("O número é perfeito?", eperfeito(int(input())))

File: 29-04/eperfeito.hs
funcMod a b
    | a == b = 0
    | a > b = funcMod (a - b) b
    | a < b = a

somaDiv n d s
    | d == 0 = s
    | funcMod n d == 0 = somaDiv n (d-1) (s+d)
    | otherwise = somaDiv n (d-1) s


eperfeito :: Int -> Bool
eperfeito n
    | somaDiv n (n-1) 0 == n = True
    | otherwise = False

main :: IO () 
main = do
    putStrLn "Digite um número:"
    num <- readLn
    putStrLn "O número é perfeito?"
    print (eperfeito num)